feat(core): provide message abstractions (#12)
NeuraLegion/secbox-sdk-js: packages/core/README.md

```ts
const config = new Configuration({
  cluster: 'app.neuralegion.com',
})
```

#### credentials
- type: `Credentials`

Set credentials to access the application.

```ts
const config = new Configuration({
  credentials: {
    token: 'your API key'
  }
})
```

More info about [setting up an API key](https://docs.neuralegion.com/docs/manage-your-organization#manage-organization-apicli-authentication-tokens)

#### credentialProviders
- type: `CredentialProvider[]`

Allows you to provide credentials and load it in runtime. You can pass many providers, and credentials will be loaded from the first provider which successfully provides credentials. Currently, only `EnvCredentialProvider` is accessible.

#### EnvCredentialProvider
Use this provider to read credentials from the following environment variable: `BRIGHT_TOKEN`

If the `BRIGHT_TOKEN` environment variable is not set or contains a falsy value, it will return undefined.

```ts
import { Configuration, EnvCredentialProvider } from '@secbox/core';

const credentialsProvider = new EnvCredentialProvider();
const config = new Configuration({
  credentialProviders: [credentialsProvider]
});
````

> ### Command

Reviewer: Now README contains nothing that may allow the user to start working with the package. It is necessary to explain what is the difference between `Command` and `Event`, when they should be used, how they can be used, etc

Reviewer: Please refer to LLD. Unfortunately, the examples below are poor and they are not able to give an understanding of the whole picture.

You should avoid superficial descriptions for each entity that is exposed by the package. 

See how it is done in nest.js.

e.g.

### Request-response

What is the command? 

What should be done to create a custom command?

```
Example 1
```
Which options can be used to adjust its behavior?

| Option | Description |
|-|-|
| name  | desc |

### Publish-subscribe

What is the event? 

What should be done to create a custom event?
```
Example 1
```

Which options can be used to adjust its behavior?

| Option | Description |
|-|-|
| name  | desc |

### Sending messages

How to execute a custom command?

```
Example 2
```

### Publishing events

How to fire a custom event?

```
Example 2
```

`abstract class` which one should extend your command class.
```ts
export class Test<T, R> extend Command<T, R> {
  // ...
}
```

The Command can be executed in a way that is more approach you or convenient from the client's perspective.
```ts
await new Command<T, R>(/*parameters*/).execute(dispatcher);

> await dispatcher.execute(await new Command<T, R>(/*parameters*/));

Reviewer: ```suggestion
await dispatcher.execute(new Command<T, R>(/*parameters*/));
```

> await dispatcher.execute(await new Command<T, R>(/*parameters*/));

Reviewer: Please don't provide examples that won't work at all 😄

```

Command can be execute synchroniouse, to do it you should set `expectReply` to `true`.

By default in `Command` presented the next options:
- `payload`
- `expectReply`
- `ttl`
- `type`
- `correlationId`

> - `createdAt`

Reviewer: Such enumeration does not make any sense. I suggest you either use Toc or just remove it

#### payload (`required`)
- type depend from command

> Data that should be sent in queue

Reviewer: What if we use Kafka or Redis streams? The description is too specific.

#### expectReply (`optional`)
- type `boolean`

Indicates whether to wait for a response. By default `true`.

#### ttl (`optional`)
- type `number`

Time what we should wait response. By default `number`.

#### type (`optional`)
- type 'string'

The name of payload type. Will be taken `payload` constructor name

#### corelationId  (`optional`)
- type `string`

Id that used to join response to the command. By default random uuid.

#### createdAt (`optional`)
- type `Date`

> The date when command instans was created. By default curent time.

Reviewer: Use a table instead

e.g.

| Props                         | Description                                                                                                                            |
| ----------------------------- | -------------------------------------------------------------------------------------------------------------------------------------- |
| `command.payload`                | Set the payload to be sent to the handler.                |
| `command.type`           | Set the type of command. By default, it equals the name of the derived command                                     |

# @secbox/core

> The core package can be used to obtain a config including credentials from different sources, and provide a simplified abstraction to handle events and commands.

Reviewer: Please update README accordingly, adding info about commands and events

});
```

### Command

> `abstract class` which one should extend your command class.

Reviewer: Don't use '`' to highlight phrases, it should be adopted to format code or highlight keywords

```ts
export class Test<T, R> extend Command<T, R> {
  // ...
}
```

The Command can be executed in a way that is more approach you or convenient from the client's perspective.

```ts
await new Command<T, R>(/*parameters*/).execute(dispatcher);
await dispatcher.execute(new Command<T, R>(/*parameters*/));
```

`Command` should be used in the case when you need execute NeuraLegion application command. For example registred repeater.

```ts
export class RegisterRepeater {
  constructor(
    public readonly repeaterId: string,
    public readonly version: string,
    public readonly localScriptsUsed: boolean
  ) {}
}

await new Command(new RegisterRepeater(repeaterId, version, false)).execute(
  dispatcher
);

> ```

Reviewer: It seems you misunderstand the main concept. Please refer to LLD

Command can be execute synchroniouse, to do it you should set `expectReply` to `true`.

|           Option           |   Type    | required/optional | Description                                                           |

> | :------------------------: | :-------: | :---------------: | --------------------------------------------------------------------- |

Reviewer: Get rid of `Type` and `required/optional` columns, they are useless

> |   **_command.payload_**    | dynamick  |    _required_     | Data that will be passed, by EventBus                                 |

Reviewer: Why do you use so much formatting here?
```suggestion
|   `command.payload`    | dynamick  |    _required_     | Message that we want to transmit to the remote service.                                 |
```

> | **_command.expectReply_**  | _boolean_ |    _optional_     | Indicates whether to wait for a response. By default `true`.          |

Reviewer: ```suggestion
| **_command.expectReply_**  | _boolean_ |    _optional_     | Indicates whether to wait for a reply. By default `true`.          |
```

> |     **_command.ttl_**      | _number_  |    _optional_     | Time what we should wait response. By default `10000`.                |

Reviewer: ```suggestion
|     `command.ttl`      | _number_  |    _optional_     | Period of time that command should be handled before being discarded. By default `10000` ms.                |
```

> |     **_command.type_**     | _string_  |    _optional_     | The name of payload type. Will be taken `payload` constructor name    |

Reviewer: As I mentioned above that has nothing to do with payload. The type is the command name.

Author: The type column was remove

> | **_command.corelationId_** | _string_  |    _optional_     | Id that used to join response to the command. By default random uuid. |

Reviewer: As I mentioned above that you should avoid such implementation details. It might be used to "join" a reply to a particular command, but it depends on implementation.

> | **_command.corelationId_** | _string_  |    _optional_     | Id that used to join response to the command. By default random uuid. |

Reviewer: ```suggestion
| **_command.corelationId_** | _string_  |    _optional_     | ID that used to join response to the command. |
```

> |  **_command.createdAt_**   |  _Date_   |    _optional_     | The date when command instans was created. By default curent time.    |

Reviewer: ```suggestion
|  **_command.createdAt_**   |  _Date_   |    _optional_     | The exact date and time the command was created.  |
```
